Mathlib/Testing/SlimCheck/Sampleable.lean

```diff
@@ -0,0 +1,324 @@
+
+## Tags
+
+random testing
+
+## References
+  * https://hackage.haskell.org/package/QuickCheck
+-/
+
+/-- This structure allows us to hide functions (especially noncomputable
+ones) in `Prop`, thus not requiring the compiler to generate any code
+for them. This is interesting if you have (noncomputable) functions
+you want to store in datatypes for proofs but not have generated any
+code for since they are not used.
+-/
+structure PhantomFunction (α : Sort u) (β : Sort v) where
+  f : α → β → Prop
+  total : ∀ x, ∃ y, f x y
+  determ : ∀ x y0 y1, f x y0 → f x y1 → y0 = y1
+
+namespace PhantomFunction
+
+/-- Evaluate the function hidden inside `ph`. -/
+noncomputable def eval (ph : PhantomFunction α β) (x : α) : β := Classical.choose $ ph.total x
+
+/-- Evaluating the function hidden inside a `PhantomFunction` is equivalent
+to evaluating the original function. -/
+@[simp]
+theorem eval_eq (ph : PhantomFunction α β) : ∀ (x : α) (y : β), ph.eval x = y ↔ ph.f x y := by
+  intro x y
+  simp [eval]
+  have h' := Classical.choose_spec <| ph.total x
+  constructor <;> intro h
+  next =>
+    rw [h] at h'
+    assumption
+  next =>
+    apply ph.determ x
+    exact h'
+    exact h
+
+/-- Hide a regular function inside a `PhantomFunction`. -/
+def mkPhantomFunction (g : α → β) : PhantomFunction α β where
+  f x y := g x = y
+  total x := ⟨_, rfl⟩
+  determ x y0 y1 h0 h1 := by rw [←h0, ←h1]
+
+/-- A specialized `mkPhantomFunction` for `SizeOf` instances. -/
+def mkSizeOf (α : Type u) [inst : SizeOf α] : PhantomFunction α Nat :=
+  mkPhantomFunction inst.sizeOf
+
+/-- A typeclass to contain `SizeOf` instances hidden inside a `PhantomFunction`
+unlike the auto generated `SizeOf` instances this one can be stored
+in datatypes for proof purposes. -/
+class PhantomSizeOf (α : Sort u) where
+  phSizeOf : PhantomFunction α Nat
+
+instance phantomSizeOf [SizeOf α] : PhantomSizeOf α where
+  phSizeOf := mkSizeOf α
+
+noncomputable instance sizeOfPhantom [ph : PhantomSizeOf α] : SizeOf α where
+  sizeOf := ph.phSizeOf.eval
+
+/-- A specialized `eval_eq` for `SizeOf` instances. -/
+@[simp]
+theorem sizeOfPhantom_eq_SizeOf [inst : SizeOf α] : @sizeOfPhantom α (@phantomSizeOf α inst) = inst := by
+  cases inst
+  unfold sizeOfPhantom
+  apply congrArg
+  funext x
+  rw [eval_eq]
+  rfl
+
+end PhantomFunction
+
+namespace SlimCheck
+
+open Random Gen PhantomFunction
+
+/-- `ShrinkFn α` is the type of functions that shrink an argument of type `α` -/
+abbrev ShrinkFn (α : Type u) [PhantomSizeOf α] := (x : α) → List { y : α // sizeOfPhantom.sizeOf y < sizeOfPhantom.sizeOf x }
+
+/-- `Sampleable α` provides ways of creating examples of type `α`. -/
+class Sampleable (α : Type u) [Repr α] where
+  sample {} : Gen α
+
+/-- Given an example `x : α`, `Shrinkable α` gives us a way to shrink it
+and suggest simpler examples. -/
+class Shrinkable (α : Type u) [phSz : PhantomSizeOf α] where
+
+  shrink : @ShrinkFn α phSz := λ _ => []
```

Comment:
```suggestion
class Shrinkable (α : Type u) extends WellFoundedRelation α where
  shrink : ShrinkFn α := λ _ => []

attribute [instance low-1] Shrinkable.toWellFoundedRelation

```

Given that SizeOf/WellFoundedRelation instances have so many non-commuting diamonds, it's probably better to bundle them.  (Otherwise whenever you write `[Shrinkable (Nat × Nat)]`, Lean requires that `PhantomSizeOf (Nat × Nat)` produces exactly the same instance as it did in the instance declaration of `Shrinkable (α × β)`.  And these can easily be different due to all the overlapping instances.)